sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoCoder.java

```diff
@@ -269,21 +278,61 @@ public ExtensionRegistry getExtensionRegistry() {
   private transient ExtensionRegistry memoizedExtensionRegistry;
   private transient Parser<T> memoizedParser;
 
+  // Descriptor used by DynamicMessage.
+  private transient Descriptors.Descriptor protoMessageDescriptor;
+
   /** Private constructor. */
   private ProtoCoder(Class<T> protoMessageClass, Set<Class<?>> extensionHostClasses) {
     this.protoMessageClass = protoMessageClass;
     this.extensionHostClasses = extensionHostClasses;
+    this.protoMessageDescriptor = null;
+  }
+
+  private ProtoCoder(
+      Descriptors.Descriptor protoMessageDescriptor, Set<Class<?>> extensionHostClasses) {
+    @SuppressWarnings("unchecked")
+    Class<T> protoMessageClass = (Class<T>) DynamicMessage.class;
+    this.protoMessageClass = protoMessageClass;
+    this.extensionHostClasses = extensionHostClasses;
+    this.protoMessageDescriptor = protoMessageDescriptor;
+  }
+
+  private void writeObject(ObjectOutputStream oos) throws IOException {
```

Comment:
Hmm, interesting. It is a problem that ProtoCoder is a CustomCoder serialized via Java serialization rather than a StructuredCoder. Not related to your change, just noting. This essentially makes it incompatible with pipeline update. @reuvenlax

Reply:
I think I pulled the upgradability of by using (and making protoMessageClass transient):
```    
oos.defaultWriteObject();
if (DynamicMessage.class.equals(this.protoMessageClass)) {
```
I don't know if a StructuredCoder would help because a Descriptor is quite a complex beast (I serialize the complete proto binary on the stream.

Reply:
Oof, that's pretty bad. Maybe we should change that?

An alternative would be to just put the DynamicMessage support in a ProtoSchemaProvider, and encourage people to stop using ProtoCoder altogether.

Reply:
Well, | don't see another way todo this. I need to get the Descriptor to the workers. The only way I see it working is sending the FileDescriptorSet as a proto steam onto the Java serialized object stream.

Getting everyone on ProtoSchemaProvider will have the exact same problem I'm afraid. I'll know in a few days when I got a working prototype.

Reply:
Could this be cleaner if you had `class DynamicMessageProtoCoder extends ProtoCoder`? Then you wouldn't need to branch based on `protoMessageClass == DynamicMessage` you just override the relevant functions.

Reply:
Wondering the same thing. Might be cleaner to have a separate Coder for this case.

Reply:
Split the DynamicProtoCoder from the ProtoCoder